tests/e2e/test_read.py

```diff
@@ -39,29 +44,87 @@ def container_client(account_url):
     container.delete_container()
 
 
-@pytest.fixture
-def blob_client(container_client, sample_data):
-    blob_name = random_resource_name()
-    blob_client = container_client.get_blob_client(blob=blob_name)
-    blob_client.upload_blob(sample_data)
-    return blob_client
+@pytest.fixture(scope="module")
+def small_blob(account_url, container_client):
+    return upload_blob(account_url, container_client, sample_data(20))
 
 
-@pytest.fixture
-def blob_url(account_url, container_client, blob_client):
-    return f"{account_url}/{container_client.container_name}/{blob_client.blob_name}"
+@pytest.fixture(scope="module")
+def large_blob(account_url, container_client):
+    return upload_blob(
+        account_url, container_client, sample_data(_PARTITIONED_DOWNLOAD_THRESHOLD * 2)
+    )
+
+
+@pytest.fixture(scope="module")
+def small_with_newlines_blob(account_url, container_client):
+    return upload_blob(account_url, container_client, sample_data_with_newlines(20, 2))
 
 
 @pytest.fixture
-def blob_io(blob_url):
-    yield BlobIO(blob_url=blob_url, mode="rb")
+def blob(request):
+    return request.getfixturevalue(f"{request.param}_blob")
 
 
 def random_resource_name(name_length=8):
-    return "".join(random.choices(string.ascii_lowercase + string.digits, k=name_length))
+    return "".join(
+        random.choices(string.ascii_lowercase + string.digits, k=name_length)
+    )
+
+
+def sample_data(data_length=20):
+    return os.urandom(data_length)
+
+
+def sample_data_with_newlines(data_length=20, num_lines=1):
+    lines = []
+    for i in range(num_lines):
+        lines.append(sample_data(int(data_length / num_lines)))
+    return b"\n".join(lines)
+
+
+def upload_blob(account_url, container_client, data):
+    blob_name = random_resource_name()
+    blob_client = container_client.get_blob_client(blob=blob_name)
+    blob_client.upload_blob(data)
+    url = f"{account_url}/{container_client.container_name}/{blob_name}"
+
+    return Blob(data=data, url=url)
 
 
 class TestRead:
-    def test_reads_all_data(self, blob_io, sample_data):
-        with blob_io as f:
-            assert f.read() == sample_data
+    @pytest.mark.parametrize(
+        "blob",
+        [
+            "small",
+            "large",
+        ],
+        indirect=True,
+    )
+    def test_reads_all_data(self, blob):
+        with BlobIO(blob.url, "rb") as f:
+            assert f.read() == blob.data
+            assert f.tell() == len(blob.data)
+
+    @pytest.mark.parametrize("n", [1, 5, 20, 21])
+    def test_read_n_bytes(self, small_blob, n):
+        with BlobIO(small_blob.url, "rb") as f:
+            for i in range(0, len(small_blob.data), n):
+                assert f.read(n) == small_blob.data[i : i + n]
+                if n < len(small_blob.data):
```

Comment:
Let's avoid if statements related to assertion making; it can cause unintentional gaps in our testing if the if statement does not behave as expected. Generally, if we find ourselves needing if statements, it is a sign that we may need a separate test case. In this case, we should be able to update the assertion to something like this:
```python
                assert f.read(n) == small_blob.data[i : i + n]
                expected_position = min(i + n, len(small_blob.data))
                assert f.tell() == expected_position
```
to enforce the assertion that the tell won't go past the end of the blob.